### PR TITLE
Allow methods that require `Self:Sized` to be object safe (RFC 817)

### PR DESCRIPTION
**Pull request by nikomatsakis**
RFC 817 is not yet accepted, but I wanted to put this code up so people can see how it works. And to be ready lest it should be accepted.

cc rust-lang/rfcs#817

### CODE REVIEW

**Comment by sfackler** [comment]
Ping?

File: src/librustc_typeck/check/method/confirm.rs [hunk 1]
```diff
@@ -404,26 +404,9 @@ impl<'a,'tcx> ConfirmContext<'a,'tcx> {
...
-        // FIXME(DST). Super hack. For a method on a trait object
-        // `Trait`, the generic signature requires that
-        // `Self:Trait`. Since, for an object, we bind `Self` to the
-        // type `Trait`, this leads to an obligation
-        // `Trait:Trait`. Until such time we DST is fully implemented,
-        // that obligation is not necessarily satisfied. (In the
-        // future, it would be.) But we know that the true `Self` DOES implement
-        // the trait. So we just delete this requirement. Hack hack hack.
-        let mut method_predicates = pick.method_ty.predicates.instantiate(self.tcx(), &all_substs);
-        match pick.kind {
```
Comment (huonw) [inline src/librustc_typeck/check/method/confirm.rs]: It is weird to me that this code can just be plain deleted, was it serving no purpose?

**Comment by nikomatsakis** [comment]
@huonw that code was serving a purpose when `Trait : Trait` was not always true. It serves no purpose now.

**Comment by nikomatsakis** [comment]
@huonw just rebased and added the code about by-value-self, should be ready for r? now

**Pull request merged**